File: api/attendance/today-status.php
<?php

require_once $_SERVER['DOCUMENT_ROOT'] . '/includes/session_start.php';
header('Content-Type: application/json; charset=utf-8');
header('Cache-Control: no-cache, no-store, must-revalidate');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

date_default_timezone_set('Asia/Tehran');

// به جای تعریف مجدد PDO
require_once $_SERVER['DOCUMENT_ROOT'] . '/config/database.php';
require_once $_SERVER['DOCUMENT_ROOT'] . '/includes/error_config.php';

try {
    $database = new Database();
    $db = $database->getConnection();
} catch (Exception $e) {
    http_response_code(500);
    echo json_encode(['success' => false, 'message' => 'Database connection error']);
    exit;
}

// ✅ استفاده از کلاس Auth برای validate کردن JWT
require_once $_SERVER['DOCUMENT_ROOT'] . '/includes/auth.php';
$auth = new Auth($db);

$user_id = null;

// روش 1: SESSION
if (isset($_SESSION['user_id'])) {
    $user_id = $_SESSION['user_id'];
}

// روش 2: JWT Token
if (!$user_id) {
    $user_id = $auth->getUserFromToken();
}

if (!$user_id) {
    http_response_code(401);
    echo json_encode([
        'success' => false,
        'message' => 'Unauthorized - No valid session or token'
    ]);
    exit;
}

// ============================================
// دریافت وضعیت حضور
// ============================================

try {
    $today = date('Y-m-d');
    
    // دریافت اطلاعات کاربر
    $stmt = $db->prepare("
        SELECT 
            shift_count,
            shift_1_start,
            shift_1_end,
            shift_2_start,
            shift_2_end,
            monthly_salary,
            daily_work_hours
        FROM users 
        WHERE id = ?
    ");
    $stmt->execute([$user_id]);
    $user = $stmt->fetch(PDO::FETCH_ASSOC);
    
    if (!$user) {
        throw new Exception('User not found');
    }
    
    // دریافت رکوردهای حضور امروز
    $stmt = $db->prepare("
        SELECT 
            id,
            shift_number,
            check_in,
            check_out
        FROM attendance_records
        WHERE user_id = ? AND date = ?
        ORDER BY shift_number ASC
    ");
    $stmt->execute([$user_id, $today]);
    $records = $stmt->fetchAll(PDO::FETCH_ASSOC);
    
    // پردازش وضعیت
    $shift1_check_in = null;
    $shift1_check_out = null;
    $shift2_check_in = null;
    $shift2_check_out = null;
    
    foreach ($records as $record) {
        if ($record['shift_number'] == 1) {
            $shift1_check_in = $record['check_in'];
            $shift1_check_out = $record['check_out'];
        } elseif ($record['shift_number'] == 2) {
            $shift2_check_in = $record['check_in'];
            $shift2_check_out = $record['check_out'];
        }
    }
    
    // تعیین دکمه‌ها
    $buttons = [];
    $window_message = null; // پیامِ «الان زمانِ ثبت ورود نیست»

    if ($user['shift_count'] == 1) {
        if (!$shift1_check_in) {
            $buttons[] = [
                'type' => 'check_in',
                'shift' => 1,
                'label' => 'ورود',
                'enabled' => true
            ];
        } elseif (!$shift1_check_out) {
            $buttons[] = [
                'type' => 'check_out',
                'shift' => 1,
                'label' => 'خروج',
                'enabled' => true
            ];
        }
    } else {
        // ساعتِ فعلی برای بررسیِ پنجره‌های مجاز
        $now_obj = new DateTime(date('H:i:s'));
        $shift1_end_obj = !empty($user['shift_1_end']) ? new DateTime($user['shift_1_end']) : null;
        // شروعِ پنجره‌ی شیفت ۲ = ۳۰ دقیقه قبل از شروعِ شیفت ۲
        $shift2_allowed_obj = null;
        if (!empty($user['shift_2_start'])) {
            $shift2_allowed_obj = new DateTime($user['shift_2_start']);
            $shift2_allowed_obj->modify('-30 minutes');
        }

        // آیا ورود شیفت ۱ هنوز مجاز است؟ (قبل از پایانِ شیفت ۱)
        $shift1_in_open = ($shift1_end_obj === null) ? true : ($now_obj < $shift1_end_obj);
        // آیا ورود شیفت ۲ مجاز است؟ (بعد از ۳۰ دقیقه قبلِ شیفت ۲)
        $shift2_in_open = ($shift2_allowed_obj === null) ? true : ($now_obj >= $shift2_allowed_obj);

        // 🔒 اول بررسی می‌کنیم آیا شیفتِ ۲ همین الان «باز» است (ورود ثبت شده، خروج نه) —
        // مستقل از اینکه شیفتِ ۱ اصلاً زده شده یا نه. قبلاً این حالت چک نمی‌شد: وقتی
        // کاربر پنجره‌ی شیفتِ ۱ را کامل از دست می‌داد و مستقیم وارد شیفتِ ۲ می‌شد،
        // شرطِ زیر (elseif زنجیره‌ای بر پایه‌ی shift1_check_in) هیچ‌وقت به بخشِ
        // «خروجِ شیفتِ ۲» نمی‌رسید و همان دکمه‌ی سبزِ «ورود شیفت ۲» دوباره نمایش
        // داده می‌شد — یعنی بعد از ثبتِ ورودِ شیفتِ ۲، دکمه هرگز قرمز/خروج نمی‌شد.
        if ($shift2_check_in && !$shift2_check_out) {
            $buttons[] = ['type' => 'check_out', 'shift' => 2, 'label' => 'خروج شیفت 2', 'enabled' => true];
        } elseif ($shift2_check_in && $shift2_check_out) {
            // هر دو شیفت کامل شده (چه شیفتِ ۱ زده شده باشه چه رد شده باشه) — بدونِ دکمه
        } elseif (!$shift1_check_in) {
            // هنوز ورود شیفت ۱ نخورده
            if ($shift1_in_open) {
                $buttons[] = ['type' => 'check_in', 'shift' => 1, 'label' => 'ورود شیفت 1', 'enabled' => true];
            } elseif ($shift2_in_open) {
                // پنجره‌ی شیفت ۱ بسته شده ⟵ مستقیم ورود شیفت ۲
                $buttons[] = ['type' => 'check_in', 'shift' => 2, 'label' => 'ورود شیفت 2', 'enabled' => true];
            } else {
                // بینِ دو شیفت ⟵ هیچ ورودی مجاز نیست
                $window_message = 'الان زمانِ ثبتِ ورود نیست';
            }
        } elseif (!$shift1_check_out) {
            // شیفت ۱ باز است ⟵ همیشه می‌تواند خروجِ شیفت ۱ بزند
            $buttons[] = ['type' => 'check_out', 'shift' => 1, 'label' => 'خروج شیفت 1', 'enabled' => true];
        } else {
            // شیفت ۱ تمام شده، شیفتِ ۲ هنوز شروع نشده ⟵ ورود شیفت ۲ فقط در پنجره‌ی مجاز
            if ($shift2_in_open) {
                $buttons[] = ['type' => 'check_in', 'shift' => 2, 'label' => 'ورود شیفت 2', 'enabled' => true];
            } else {
                $window_message = 'الان زمانِ ثبتِ ورود نیست';
            }
        }
    }
    
    // پیام وضعیت
    $status_message = '';
    if ($shift1_check_in) {
        $status_message = 'ورود ثبت شد: ' . substr($shift1_check_in, 11, 5);
        if ($shift1_check_out) {
            $status_message .= ' | خروج: ' . substr($shift1_check_out, 11, 5);
        }
    }
    if ($shift2_check_in) {
        $status_message .= ' | ورود ش۲: ' . substr($shift2_check_in, 11, 5);
        if ($shift2_check_out) {
            $status_message .= ' | خروج ش۲: ' . substr($shift2_check_out, 11, 5);
        }
    }
    
    http_response_code(200);
    echo json_encode([
        'success' => true,
        'shift_count' => (int)$user['shift_count'],
        'buttons' => $buttons,
        'window_message' => $window_message,
        'status_message' => $status_message,
        'current_time' => date('H:i'),
        'shift1' => [
            'check_in' => $shift1_check_in ? substr($shift1_check_in, 11, 5) : null,
            'check_out' => $shift1_check_out ? substr($shift1_check_out, 11, 5) : null
        ],
        'shift2' => [
            'check_in' => $shift2_check_in ? substr($shift2_check_in, 11, 5) : null,
            'check_out' => $shift2_check_out ? substr($shift2_check_out, 11, 5) : null
        ]
    ]);
    
} catch (Exception $e) {
    http_response_code(500);
    error_log("Today status error: " . $e->getMessage());
    echo json_encode([
        'success' => false,
        'message' => 'خطا: ' . $e->getMessage()
    ]);
}
?>

File: go-api/internal/attendance/today_status.go
package attendance

import (
	"database/sql"
	"net/http"
	"time"

	"bmp/go-api/internal/core"
)

// parseClock — یک رشتهٔ ساعتِ MySQL ("HH:MM:SS" یا "HH:MM") را روی یک
// تاریخِ ثابتِ دلخواه (برایِ مقایسه/جمع‌وتفریق) پارس می‌کند. معادلِ
// `new DateTime($timeString)` در PHP وقتی فقط رشتهٔ ساعت داده می‌شود
// (PHP خودش تاریخِ امروز را فرض می‌کند؛ اینجا یک تاریخِ ثابتِ دلخواه کافی
// است چون فقط برایِ مقایسهٔ نسبی استفاده می‌شود).
func parseClock(s string) (time.Time, bool) {
	if s == "" {
		return time.Time{}, false
	}
	for _, layout := range []string{"15:04:05", "15:04"} {
		if t, err := time.Parse(layout, s); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

type statusButton struct {
	Type    string `json:"type"`
	Shift   int    `json:"shift"`
	Label   string `json:"label"`
	Enabled bool   `json:"enabled"`
}

// TodayStatus — پورتِ دقیقِ api/attendance/today-status.php
//
//	GET /go/api/attendance/today-status
//	→ {"success":true,"shift_count":N,"buttons":[...],"window_message":str|null,
//	   "status_message":str,"current_time":"HH:MM","shift1":{...},"shift2":{...}}
func TodayStatus(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		u := core.UserOf(r.Context())
		now := core.TehranNow()
		today := now.Format("2006-01-02")

		var shiftCount sql.NullInt64
		var shift1Start, shift1End, shift2Start, shift2End sql.NullString
		err := db.QueryRow(`
			SELECT shift_count, shift_1_start, shift_1_end, shift_2_start, shift_2_end
			FROM users WHERE id = ?
		`, u.ID).Scan(&shiftCount, &shift1Start, &shift1End, &shift2Start, &shift2End)
		if err == sql.ErrNoRows {
			core.WriteErr(w, http.StatusInternalServerError, "خطا در پردازش درخواست")
			return
		}
		if err != nil {
			core.WriteErr(w, http.StatusInternalServerError, "خطا در پردازش درخواست")
			return
		}

		rows, err := db.Query(`
			SELECT shift_number, check_in, check_out
			FROM attendance_records
			WHERE user_id = ? AND date = ?
			ORDER BY shift_number ASC
		`, u.ID, today)
		if err != nil {
			core.WriteErr(w, http.StatusInternalServerError, "خطا در پردازش درخواست")
			return
		}
		defer rows.Close()

		var shift1In, shift1Out, shift2In, shift2Out sql.NullString
		for rows.Next() {
			var shiftNum int
			var checkIn, checkOut sql.NullString
			if err := rows.Scan(&shiftNum, &checkIn, &checkOut); err != nil {
				core.WriteErr(w, http.StatusInternalServerError, "خطا در پردازش درخواست")
				return
			}
			if shiftNum == 1 {
				shift1In, shift1Out = checkIn, checkOut
			} else if shiftNum == 2 {
				shift2In, shift2Out = checkIn, checkOut
			}
		}

		buttons := []statusButton{}
		var windowMessage any = nil

		if shiftCount.Int64 == 1 {
			if !shift1In.Valid {
				buttons = append(buttons, statusButton{"check_in", 1, "ورود", true})
			} else if !shift1Out.Valid {
				buttons = append(buttons, statusButton{"check_out", 1, "خروج", true})
			}
		} else {
			nowClock, _ := parseClock(now.Format("15:04:05"))

			shift1InOpen := true
			if shift1End.Valid && shift1End.String != "" {
				if end, ok := parseClock(shift1End.String); ok {
					shift1InOpen = nowClock.Before(end)
				}
			}

			shift2InOpen := true
			if shift2Start.Valid && shift2Start.String != "" {
				if start, ok := parseClock(shift2Start.String); ok {
					allowed := start.Add(-30 * time.Minute)
					shift2InOpen = !nowClock.Before(allowed)
				}
			}

			switch {
			case shift2In.Valid && !shift2Out.Valid:
				buttons = append(buttons, statusButton{"check_out", 2, "خروج شیفت 2", true})
			case shift2In.Valid && shift2Out.Valid:
				// هر دو شیفت کامل شده — بدونِ دکمه
			case !shift1In.Valid:
				if shift1InOpen {
					buttons = append(buttons, statusButton{"check_in", 1, "ورود شیفت 1", true})
				} else if shift2InOpen {
					buttons = append(buttons, statusButton{"check_in", 2, "ورود شیفت 2", true})
				} else {
					windowMessage = "الان زمان ثبت ورود نیست"
				}
			case !shift1Out.Valid:
				buttons = append(buttons, statusButton{"check_out", 1, "خروج شیفت 1", true})
			default:
				if shift2InOpen {
					buttons = append(buttons, statusButton{"check_in", 2, "ورود شیفت 2", true})
				} else {
					windowMessage = "الان زمان ثبت ورود نیست"
				}
			}
		}

		statusMessage := ""
		if shift1In.Valid {
			statusMessage = "ورود ثبت شد: " + hhmm(shift1In.String)
			if shift1Out.Valid {
				statusMessage += " | خروج: " + hhmm(shift1Out.String)
			}
		}
		if shift2In.Valid {
			statusMessage += " | ورود ش۲: " + hhmm(shift2In.String)
			if shift2Out.Valid {
				statusMessage += " | خروج ش۲: " + hhmm(shift2Out.String)
			}
		}

		core.WriteJSON(w, http.StatusOK, map[string]any{
			"success":        true,
			"shift_count":    int(shiftCount.Int64),
			"buttons":        buttons,
			"window_message": windowMessage,
			"status_message": statusMessage,
			"current_time":   now.Format("15:04"),
			"shift1": map[string]any{
				"check_in":  nsHHMM(shift1In),
				"check_out": nsHHMM(shift1Out),
			},
			"shift2": map[string]any{
				"check_in":  nsHHMM(shift2In),
				"check_out": nsHHMM(shift2Out),
			},
		})
	}
}

// hhmm — معادلِ substr($datetime, 11, 5) در PHP: از یک رشتهٔ
// "YYYY-MM-DD HH:MM:SS" فقط بخشِ "HH:MM" را برمی‌دارد.
func hhmm(s string) string {
	if len(s) >= 16 {
		return s[11:16]
	}
	return s
}

func nsHHMM(n sql.NullString) any {
	if !n.Valid {
		return nil
	}
	return hhmm(n.String)
}
